src/FreeAge/server/game.cpp

```diff
@@ -1352,16 +1396,86 @@ void Game::ProduceUnit(ServerBuilding* building, UnitType unitInProduction) {
   u32 newUnitId;
   ServerUnit* newUnit = map->AddUnit(building->GetPlayerIndex(), unitInProduction, QPointF(-999, -999), &newUnitId);
   
+  QPointF freeSpace;
+  bool foundFreeSpace = FindFreeSpaceAroundBuilding(building, newUnit, freeSpace);
+  
+  if (foundFreeSpace) {
+    newUnit->SetMapCoord(freeSpace);
+  } else {
+    // TODO (maanoo): Garrison the unit in the building
+  }
+  
+  // Send messages to clients that see the new unit
+  QByteArray addObjectMsg = CreateAddObjectMessage(newUnitId, newUnit);
+  for (auto& player : *playersInGame) {
+    accumulatedMessages[player->index] += addObjectMsg;
+  }
+  
+  GetPlayerStats(newUnit->GetPlayerIndex())->UnitAdded(unitInProduction);
+}
+
+void Game::GarrisonUnit(u32 unitId, ServerUnit* unit, u32 targetObjectId, ServerObject* targetObject, bool enter) {
+  // TODO: implement a way to get ID from ClientObject, then remove the ids for the arguments
+  if (enter) {
+    targetObject->GarrisonUnit(unitId);
+    unit->SetGarrisonedInsideObject(targetObjectId);
+    unit->StopMovement();
+    unit->RemoveTarget();
+
+    QByteArray unitGarrisonMessage = CreateUnitGarrisonMessage(unitId, targetObjectId);
+    QByteArray unitMoveMessage = CreateUnitMovementMessage(unitId,
+      unit->GetMapCoord(),
+      unit->GetMoveSpeed() * unit->GetMovementDirection(),
+      unit->GetCurrentAction());
+    for (auto& player : *playersInGame) {
+      accumulatedMessages[player->index] += unitGarrisonMessage;
+      accumulatedMessages[player->index] += unitMoveMessage;
+    }
+  } else {
+    if (!targetObject->isBuilding()) {
+      // TODO: implement ungarrison from unit
+      LOG(WARNING) << "Ungarrison from unit not implemented yet";
+      return;
+    }
+    ServerBuilding* targetBuilding = AsBuilding(targetObject);
+    if (QPointF freeSpace; FindFreeSpaceAroundBuilding(targetBuilding, unit, freeSpace)) {
+      targetBuilding->UngarrisonUnit(unitId);
+      unit->SetGarrisonedInsideObject(kInvalidObjectId);
+      unit->SetMapCoord(freeSpace);
+      unit->StopMovement();
```

Comment:
If I don't call the StopMovement weird things happen...

When the unit gets in a building I call this:
```
    targetObject->GarrisonUnit(unitId);
    unit->SetGarrisonedInsideObject(targetObjectId);
    unit->StopMovement();
    unit->RemoveTarget();
```
Am I forgetting anythin?

Reply:
I hope not, but I am not 100% sure either. This is clearly a sign that the ServerUnit interface / handling needs both better documentation, and if possible, a redesign to make it less error-prone.

Reply:
I found the source of the weird thing: `overrideDirection`